src/rate_limiter.cc

```diff
@@ -775,52 +757,89 @@ RateLimiter::ResourceManager::AddModelInstance(
           resource.count();
     }
   }
+  // Increase max resource if needed.
+  std::lock_guard<std::mutex> lk2(max_resources_mtx_);
+  UpdateMaxResource(pr.first->second);
+  return ParseAndValidateResources();
 }
 
 Status
 RateLimiter::ResourceManager::RemoveModelInstance(
     const ModelInstanceContext* instance)
 {
-  std::lock_guard<std::mutex> lk(model_resources_mtx_);
+  // Find instance from model resources.
+  std::lock_guard<std::mutex> lk1(model_resources_mtx_);
   const auto& itr = model_resources_.find(instance);
   if (itr == model_resources_.end()) {
     return Status(
         Status::Code::INTERNAL, "Can not find the instance to remove");
   }
+  // Check if max resources need to be updated.
+  bool update_needed = false;
+  std::lock_guard<std::mutex> lk2(max_resources_mtx_);
+  for (const auto& resource_device_map : itr->second) {
+    auto ditr = max_resources_.find(resource_device_map.first);
+    if (ditr != max_resources_.end()) {
+      for (const auto resource : resource_device_map.second) {
+        auto ritr = ditr->second.find(resource.first);
+        if (ritr != ditr->second.end() && ritr->second >= resource.second) {
```

Comment:
> If the removed instance has higher or equal limit to the max resource

@kthui can an instance have a higher resource limit than the max resource?

Reply:
> can an instance have a higher resource limit than the max resource?

In the case when an instance is being removed, it cannot have higher resource limit than the max resource. For instance, there are 3 instances with resources `{1, 2, 3}`, so the current max resource is `3`. Removing any of the three instances cannot exceed the current max resource of `3`.

If the instance with the max resource is removed, the max resource is recomputed because the second to the max resource is unknown. For any other instances, the current max resource is still valid.

> Shouldn't the instance's value be considered in the current max?

In the case when it is removed, I think it does not hurt to leave it as the max, until some other instances with higher resource limit is added, but it is recomputed to match the previous behavior, which recompute the max resource after every instance add/remove.

Reply:
> In the case when it is removed, I think it does not hurt to leave it as the max
I'm not suggesting to leave it as the max when removed, I just wanted to verify my understanding that it shouldn't be possible to be higher than the max.

i.e. `ritr->second >= resource.second` may actually be more like:
```
if (ritr->second > resource.second) {
  // this shouldn't happen, bug?
} else if (ritr->second == resource.second) {
  // this instance held a max, so recompute
} else {
  // this wasn't a max, no action necessary
}
```

Reply:
Yes, you are right. The `>` case should never happen, the `>=` is only as a failsafe.

Reply:
Can we return an error status in the `>` case to catch if we ever have a bug in our resource limit logic?

Reply:
Yes, also enhanced all other "should not print this" scattered in the rate limiter: [Enhance should not print this message](https://github.com/triton-inference-server/core/pull/208/commits/0cdb4c98b1e4b2e0155f8e16cc3284698f87a1d4)

Reply:
Sorry for the confusion, my arrow is at the wrong direction 🤦: [Fix bug](https://github.com/triton-inference-server/core/pull/208/commits/ae3a3efae03e5d522a4aef2256a3a1d99fff9c61)